presto-orc/src/main/java/com/facebook/presto/orc/reader/MapFlatSelectiveStreamReader.java

```diff
@@ -0,0 +1,781 @@
+        switch (keyOrcTypeKind) {
+            case BYTE:
+            case SHORT:
+            case INT:
+            case LONG:
+                requiredLongKeys = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.LongSubscript.class::cast)
+                        .map(Subfield.LongSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredStringKeys = null;
+                return;
+            case STRING:
+            case BINARY:
+                requiredStringKeys = requiredSubfields.stream()
+                        .map(Subfield::getPath)
+                        .map(path -> path.get(0))
+                        .map(Subfield.StringSubscript.class::cast)
+                        .map(Subfield.StringSubscript::getIndex)
+                        .collect(toImmutableSet());
+                requiredLongKeys = null;
+                return;
+            default:
+                throw new IllegalArgumentException("Unsupported flat map key type: " + keyOrcTypeKind);
+        }
+    }
+
+    private static Optional<TupleDomainFilter> getTopLevelFilter(Map<Subfield, TupleDomainFilter> filters)
+    {
+        Map<Subfield, TupleDomainFilter> topLevelFilters = Maps.filterEntries(filters, entry -> entry.getKey().getPath().isEmpty());
+        if (topLevelFilters.isEmpty()) {
+            return Optional.empty();
+        }
+
+        checkArgument(topLevelFilters.size() == 1, "MAP column may have at most one top-level range filter");
+        TupleDomainFilter filter = Iterables.getOnlyElement(topLevelFilters.values());
+        checkArgument(filter == IS_NULL || filter == IS_NOT_NULL, "Top-level range filter on MAP column must be IS NULL or IS NOT NULL");
+        return Optional.of(filter);
+    }
+
+    @Override
+    public int read(int offset, int[] positions, int positionCount)
+            throws IOException
+    {
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
+
+        if (!rowGroupOpen) {
+            openRowGroup();
+        }
+
+        allNulls = false;
+
+        if (outputRequired && nullsAllowed && presentStream != null) {
+            nulls = ensureCapacity(nulls, positionCount);
+        }
+
+        if (!nullsAllowed || !nonNullsAllowed) {
+            outputPositions = ensureCapacity(outputPositions, positionCount);
+        }
+        else {
+            outputPositions = positions;
+            outputPositionsReadOnly = true;
+        }
+
+        if (presentStream != null && keyCount == 0) {
+            readAllNulls(offset, positions, positionCount);
+        }
+        else {
+            readNotAllNulls(offset, positions, positionCount);
+        }
+
+        localMemoryContext.setBytes(getRetainedSizeInBytes());
+
+        return outputPositionCount;
+    }
+
+    private void readAllNulls(int offset, int[] positions, int positionCount)
```

Comment:
We don't need `offset` if it's all nulls?

Reply:
@highker yes, `offset` is not needed if all values are `null`; I'll remove it.